eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -181,14 +201,206 @@ private void checkClassField(DetailAST objBlockToken) {
+        // Asynchronous method with annotation @ServiceMethod must not has {@code Context} as a parameter.
+        checkContextInRightPlace(methodDefToken);
+    }
+
+    /**
+     * Checks for the return type of async and sync clients should be as per guidelines:
+     *     1) The return type for async collection should be of type? extends Flux
+     *     2) The return type for async single value should be of type? extends Mono
+     *     3) The return type for sync collection should be of type? extends Stream
+     *     4) The return type for sync single value should be of type? extends Response
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     * @param serviceMethodAnnotation ANNOTATION AST node which used to find the if the annotation has 'return' key,
+     * if found. return the value of member'return'.
+     */
+    private void checkServiceClientMethodReturnType(DetailAST methodDefToken, DetailAST serviceMethodAnnotation) {
+        // Find the annotation member 'returns' value
+        String returnsAnnotationMemberValue = getAnnotationMemberReturnsValue(serviceMethodAnnotation, "returns");
+
+        StringBuilder sb = new StringBuilder();
+        getReturnType(methodDefToken.findFirstToken(TokenTypes.TYPE), sb);
+        String returnType = sb.toString();
+
+        // If value of 'returns' is SINGLE, and then log error if the return type of the method is not start with Mono or Response
+        // If value of 'returns' is COLLECTION, and then log error if the return type of the method is not start with PagedFlux or Stream
+        if (SINGLE_RETURN_TYPE.equals(returnsAnnotationMemberValue)
+            && !returnType.startsWith(MONO) && !returnType.startsWith(RESPONSE)) {
+            log(methodDefToken, String.format(SINGLE_VALUE_RETURN_ERROR, SINGLE_RETURN_TYPE));
+        } else if (COLLECTION_RETURN_TYPE.equals(returnsAnnotationMemberValue)
+            && !returnType.startsWith(PAGED_FLUX) && !returnType.startsWith(STREAM)) {
+            log(methodDefToken, String.format(COLLECTION_RETURN_ERROR, COLLECTION_RETURN_TYPE));
+        }
+    }
+
+    /**
+     * Given the method is already annotated @ServiceMethod. Checks if the return type is Response or Mono<Response>,
+     * If the return type is Response, the method name must end with 'withResponse'.
+     * If the return type is Mono<Response>. the method name must not end with 'withResponse'.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkReturnTypeNamingPattern(DetailAST methodDefToken, String methodName) {
+        DetailAST typeToken = methodDefToken.findFirstToken(TokenTypes.TYPE);
+
+        StringBuilder sb = new StringBuilder();
+        getReturnType(typeToken, sb);
+        String returnType = sb.toString();
+
+        if (returnType.startsWith(RESPONSE_GENERIC_START) && !methodName.endsWith(WITH_RESPONSE)) {
```

Comment:
This should be case sensitive. WITH_RESPONSE is declared as `withResponse` which will not work as all methods will be named `<verb>WithResponse`. So, change WITH_RESPONSE to `WithResponse` instead.